File: src/main/java/io/github/takenoha/towerdefense/paper/VoucherContainerPolicy.java
package io.github.takenoha.towerdefense.paper;

/** Directional guard for plain voucher insertion into an anvil-like inventory. */
public final class VoucherContainerPolicy {
    private VoucherContainerPolicy() {
    }

    public static boolean blocksPlainVoucherInsertion(
            boolean forbiddenInventory,
            boolean topTarget,
            boolean shiftClick,
            boolean cursorVoucher,
            boolean numberKeyVoucher,
            boolean offhandSwapVoucher,
            boolean clickedVoucher) {
        if (!forbiddenInventory) {
            return false;
        }
        if (topTarget) {
            return cursorVoucher || numberKeyVoucher || offhandSwapVoucher;
        }
        return shiftClick && clickedVoucher;
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/paper/VoucherContainerPolicy.kt
package io.github.takenoha.towerdefense.paper

/** Directional guard for plain voucher insertion into an anvil-like inventory. */
class VoucherContainerPolicy private constructor() {
    companion object {
        @JvmStatic
        fun blocksPlainVoucherInsertion(
            forbiddenInventory: Boolean,
            topTarget: Boolean,
            shiftClick: Boolean,
            cursorVoucher: Boolean,
            numberKeyVoucher: Boolean,
            offhandSwapVoucher: Boolean,
            clickedVoucher: Boolean,
        ): Boolean {
            if (!forbiddenInventory) {
                return false
            }
            return if (topTarget) {
                cursorVoucher || numberKeyVoucher || offhandSwapVoucher
            } else {
                shiftClick && clickedVoucher
            }
        }
    }
}
